File: backend/app/inn.py
"""Проверка ИНН по алгоритму контрольного числа ФНС (FR-003).

Длина зависит от организационно-правовой формы: 10 цифр у юридического
лица, 12 — у индивидуального предпринимателя, самозанятого и НКО-заявителя,
зарегистрированного как физическое лицо (ограничение chk_org_profile_inn).
"""

WEIGHTS_10 = (2, 4, 10, 3, 5, 9, 4, 6, 8)
WEIGHTS_11 = (7, 2, 4, 10, 3, 5, 9, 4, 6, 8)
WEIGHTS_12 = (3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8)

LENGTH_BY_TYPE = {"OOO": 10, "IP": 12, "NKO": 12, "SMZ": 12}


def _control_digit(digits: str, weights: tuple[int, ...]) -> int:
    return sum(int(d) * w for d, w in zip(digits, weights)) % 11 % 10


def validate_inn(inn: str, org_type: str) -> None:
    """Проверка ИНН. Возбуждает ValueError с текстом для поля формы."""
    if not inn.isdigit():
        raise ValueError("ИНН должен содержать только цифры")

    expected_length = LENGTH_BY_TYPE.get(org_type)
    if expected_length is None:
        raise ValueError("Неизвестный тип организации")
    if len(inn) != expected_length:
        raise ValueError(f"ИНН должен содержать {expected_length} цифр")

    if expected_length == 10:
        valid = _control_digit(inn[:9], WEIGHTS_10) == int(inn[9])
    else:
        valid = _control_digit(inn[:10], WEIGHTS_11) == int(inn[10]) and _control_digit(
            inn[:11], WEIGHTS_12
        ) == int(inn[11])

    if not valid:
        raise ValueError("Неверный ИНН")

File: frontend/src/lib/inn.js
/**
 * Проверка ИНН по алгоритму контрольного числа ФНС.
 *
 * Повторяет app/inn.py на сервере. Дублирование здесь осознанное: клиентская
 * проверка показывает опечатку мгновенно, серверная остаётся единственной
 * защитой, поскольку клиент можно обойти прямым запросом к API.
 */

const WEIGHTS_10 = [2, 4, 10, 3, 5, 9, 4, 6, 8];
const WEIGHTS_11 = [7, 2, 4, 10, 3, 5, 9, 4, 6, 8];
const WEIGHTS_12 = [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8];

const LENGTH_BY_TYPE = { OOO: 10, IP: 12, NKO: 12, SMZ: 12 };

function controlDigit(digits, weights) {
  const sum = weights.reduce((total, weight, index) => total + weight * Number(digits[index]), 0);
  return (sum % 11) % 10;
}

/** Возвращает текст ошибки или null, если ИНН корректен. */
export function validateInn(inn, orgType) {
  if (!/^\d+$/.test(inn)) return "ИНН должен содержать только цифры";

  const expected = LENGTH_BY_TYPE[orgType];
  if (!expected) return "Выберите тип организации";
  if (inn.length !== expected) return `ИНН должен содержать ${expected} цифр`;

  const valid =
    expected === 10
      ? controlDigit(inn, WEIGHTS_10) === Number(inn[9])
      : controlDigit(inn, WEIGHTS_11) === Number(inn[10]) &&
        controlDigit(inn, WEIGHTS_12) === Number(inn[11]);

  return valid ? null : "Неверный ИНН: контрольное число не сходится";
}
